cpu/cpu_openbsd.go

```diff
@@ -45,99 +43,64 @@ func init() {
 	if err == nil {
 		ClocksPerSec = float64(clkTck)
 	}
-
-	func() {
-		v, err := unix.Sysctl("kern.osrelease") // can't reuse host.PlatformInformation because of circular import
-		if err != nil {
-			return
-		}
-		v = strings.ToLower(v)
-		version, err := strconv.ParseFloat(v, 64)
-		if err != nil {
-			return
-		}
-		if version >= 6.4 {
-			cpIntr = 4
-			cpIdle = 5
-			cpUStates = 6
-		}
-	}()
-}
-
-func smt() (bool, error) {
-	mib := []int32{ctlHw, sMT}
-	buf, _, err := common.CallSyscall(mib)
-	if err != nil {
-		return false, err
-	}
-
-	var ret bool
-	br := bytes.NewReader(buf)
-	if err := binary.Read(br, binary.LittleEndian, &ret); err != nil {
-		return false, err
-	}
-
-	return ret, nil
 }
 
 func Times(percpu bool) ([]TimesStat, error) {
 	return TimesWithContext(context.Background(), percpu)
 }
 
-func TimesWithContext(ctx context.Context, percpu bool) ([]TimesStat, error) {
-	var ret []TimesStat
-
-	var ncpu int
-	if percpu {
-		ncpu, _ = Counts(true)
-	} else {
-		ncpu = 1
-	}
-
-	smt, err := smt()
-	if err == syscall.EOPNOTSUPP {
-		// if hw.smt is not applicable for this platform (e.g. i386),
-		// pretend it's enabled
-		smt = true
-	} else if err != nil {
-		return nil, err
+func cpsToTS(cpuTimes []uint64, name string) TimesStat {
+	return TimesStat{
+		CPU:    name,
+		User:   float64(cpuTimes[cpUser]) / ClocksPerSec,
+		Nice:   float64(cpuTimes[cpNice]) / ClocksPerSec,
+		System: float64(cpuTimes[cpSys]) / ClocksPerSec,
+		Idle:   float64(cpuTimes[cpIdle]) / ClocksPerSec,
+		Irq:    float64(cpuTimes[cpIntr]) / ClocksPerSec,
 	}
+}
 
-	for i := 0; i < ncpu; i++ {
-		j := i
-		if !smt {
-			j *= 2
-		}
-
-		cpuTimes := make([]int32, cpUStates)
-		var mib []int32
-		if percpu {
-			mib = []int32{ctlKern, kernCptime2, int32(j)}
-		} else {
-			mib = []int32{ctlKern, kernCptime}
-		}
+func TimesWithContext(ctx context.Context, percpu bool) (ret []TimesStat, err error) {
+	if !percpu {
+		mib := []int32{ctlKern, kernCpTime}
 		buf, _, err := common.CallSyscall(mib)
 		if err != nil {
 			return ret, err
 		}
+		var x []C.long
+		// could use unsafe.Slice but it's only for go1.17+
+		x = (*[cpuStates]C.long)(unsafe.Pointer(&buf[0]))[:]
```

Comment:
Sorry, It seems this PR uses CGO. we are aiming gopsutil for pure golang.

Reply:
To be honest I was tempted to rewrite the whole thing in cgo for practical reason.  By design, OpenBSD doesn't have a strong ABI compatibility, things are changed, removed an added when needed.  (This not to say that OpenBSD breaks its ABI every day, but it's not uncommon.  See the removal for the support for <6.3 for example, where various constants changed value after `hw.smt` was added.)

What OpenBSD does is to provide *source* compatibility; that is you can always recompile the software after a change.  Rewriting the constants to use values pulled from C will mean that it would be almost no breakage of this library (and the software that depends on it) due to changes on base.

IIRC Go itself on OpenBSD switched to cgo by default a couple of years ago, as now it goes thru libc to do syscalls (the same happens on solaris too.)

anyway, in this pr i'm using cgo only to know the size of `long`, which varies between platform (and OpenBSD supports a huge amount of them.)  Usually it's 64 bit on 64 bit arches and 32 on 32 bit arches, but I'm not sure and I personally own only amd64 and i686.  What do we do in these cases?

Reply:
personally I don't really like that approach -- limiting the support to i386 and amd64 for no real gain only to avoid cgo, when that's the default, the preferred way to get stuff done on this os and the only solution to avoid breakage in the future due to ABI changes.

I'll save this discussion for another thread thought, since it's the pattern already in use I'll update this pr to match existing practices in a couple of days :)

thanks